src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+fn parse_tee_evidence(report: &AttestationReport) -> Result<TeeEvidenceParsedClaim> {
+    let claims_map = json!({
+        // policy fields
+        "policy_abi_major": report.policy.abi_major(),
+        "policy_abi_minor": report.policy.api_minor(),
+        "policy_smt_allowed": report.policy.smt_allowed(),
+        "policy_migrate_ma": report.policy.migrate_ma_allowed(),
+        "policy_debug_allowed": report.policy.debug_allowed(),
+        "policy_single_socket": report.policy.single_socket_required(),
+
+        // versioning info
+        "reported_tcb_bootloader": report.reported_tcb.boot_loader,
+        "reported_tcb_tee": report.reported_tcb.tee,
+        "reported_tcb_snp": report.reported_tcb.snp,
+        "reported_tcb_microcode": report.reported_tcb.microcode,
+
+        // platform info
+        "platform_tsme_enabled": report.plat_info.tsme_enabled(),
+        "platform_smt_enabled": report.plat_info.smt_enabled(),
+
+        // measurement
+        "measurement": base64::encode(report.measurement),
+    });
+
+    Ok(claims_map as TeeEvidenceParsedClaim)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use openssl::nid::Nid;
+
+    #[test]
+    fn check_milan_certificates() {
+        let (ask, ark) = load_milan_cert_chain().unwrap();
+        assert_eq!(get_common_name(&ark).unwrap(), "ARK-Milan");
+        assert_eq!(get_common_name(&ask).unwrap(), "SEV-Milan");
+
+        assert!(ark
+            .verify(&(ark.public_key().unwrap() as PKey<Public>))
+            .map_err(|_| anyhow!("Invalid ARK Signature"))
+            .unwrap());
+
+        assert!(ask
+            .verify(&(ark.public_key().unwrap() as PKey<Public>))
+            .map_err(|_| anyhow!("Invalid ASK Signature"))
+            .unwrap());
+    }
+
+    fn get_common_name(cert: &x509::X509) -> Result<String> {
+        let mut entries = cert.subject_name().entries_by_nid(Nid::COMMONNAME);
+
+        if let Some(e) = entries.next() {
+            assert_eq!(entries.count(), 0);
+            return Ok(e.data().as_utf8()?.to_string());
+        }
+        Err(anyhow!("No CN found"))
+    }
+
+    #[test]
+    fn check_vcek_parsing() {
+        let vcek_der = include_bytes!("test-vcek.der");
+        let parsed_vcek = X509Certificate::from_der(vcek_der)
+            .unwrap()
+            .1
+            .tbs_certificate;
+
+        get_oid::<64>(&parsed_vcek, HW_ID_OID).unwrap();
+        let oids = vec![UCODE_SPL_OID, SNP_SPL_OID, TEE_SPL_OID, LOADER_SPL_OID];
+        for oid in oids {
+            get_oid_int(&parsed_vcek, oid).unwrap();
+        }
+    }
+
+    #[test]
+    fn check_vcek_signature() {
+        let (ask, _ark) = load_milan_cert_chain().unwrap();
+        let vcek = x509::X509::from_der(include_bytes!("test-vcek.der")).unwrap();
+
+        // ASK -> VCEK
+        vcek.verify(&(ask.public_key().unwrap() as PKey<Public>))
+            .map_err(|_| anyhow!("Invalid VCEK Signature"))
+            .unwrap();
+    }
```

Comment:
1. I suggest modifying this to use `verify_cert_chain()` which is the code that actually runs in the verifier:

```rust
#[cfg(test)]
mod tests {
...
    use sev::firmware::host::types::CertTableEntry;
...
    #[test]
    fn check_vcek_signature() {
        let vcek = include_bytes!("test-vcek.der").to_vec();
        let cert_table: CertTable = CertTable {
            entries: vec![
                CertTableEntry::new(SnpCertType::VCEK, vcek),
            ],
        };
        verify_cert_chain(&cert_table).unwrap();
    }
}
```

2. I suggest adding a negative check -- somehow corrupting test-vcek and checking that `verify_cert_chain()` returns an Err.

Reply:
Added